### PR TITLE
Fix mach bootstrap for Python 3.13

### LINKED ISSUE #34095: Python 3.13.0 cannot `.\mach bootstrap` properly
**Describe the bug:**
In short `.\mach bootstrap` will fail when:
- On Windows
- using Python 3.13.0

probably similar to #34022

Due to the lack of version specification in readme, I opted for the latest Python version (3.13.0) for my recent refresh of servo dev environment setup which leads to an error:
```
E:\GitHub\servo>.\mach bootstrap
 * Setting up virtual environment...
 * Upgrading pip...
 * Installing Python requirements...
Collecting blessings==1.7 (from -r E:\GitHub\servo\python\requirements.txt (line 4))
  Using cached blessings-1.7-py3-none-any.whl.metadata (19 kB)
Collecting distro==1.4 (from -r E:\GitHub\servo\python\requirements.txt (line 5))
  Using cached distro-1.4.0-py2.py3-none-any.whl.metadata (6.6 kB)
Collecting mozinfo==1.2.3 (from -r E:\GitHub\servo\python\requirements.txt (line 6))
  Using cached mozinfo-1.2.3-py2.py3-none-any.whl.metadata (649 bytes)
Collecting mozlog==8.0.0 (from -r E:\GitHub\servo\python\requirements.txt (line 7))
  Using cached mozlog-8.0.0.tar.gz (53 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting setuptools==70.0.0 (from -r E:\GitHub\servo\python\requirements.txt (line 8))
  Using cached setuptools-70.0.0-py3-none-any.whl.metadata (5.9 kB)
Collecting toml==0.9.2 (from -r E:\GitHub\servo\python\requirements.txt (line 9))
  Using cached toml-0.9.2.tar.gz (7.6 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting flake8==7.0.0 (from -r E:\GitHub\servo\python\requirements.txt (line 12))
  Using cached flake8-7.0.0-py2.py3-none-any.whl.metadata (3.8 kB)
Collecting pep8==1.5.7 (from -r E:\GitHub\servo\python\requirements.txt (line 13))
  Using cached pep8-1.5.7-py2.py3-none-any.whl.metadata (19 kB)
Collecting pyflakes==3.2.0 (from -r E:\GitHub\servo\python\requirements.txt (line 14))
  Using cached pyflakes-3.2.0-py2.py3-none-any.whl.metadata (3.5 kB)
Collecting ply==3.8 (from -r E:\GitHub\servo\python\requirements.txt (line 17))
  Using cached ply-3.8.tar.gz (157 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting colorama==0.3.7 (from -r E:\GitHub\servo\python\requirements.txt (line 20))
  Using cached colorama-0.3.7-py2.py3-none-any.whl.metadata (13 kB)
Collecting boto3==1.34.95 (from -r E:\GitHub\servo\python\requirements.txt (line 23))
  Using cached boto3-1.34.95-py3-none-any.whl.metadata (6.6 kB)
Collecting pyOpenSSL==23.0.0 (from -r E:\GitHub\servo\python\requirements.txt (line 24))
  Using cached pyOpenSSL-23.0.0-py3-none-any.whl.metadata (9.5 kB)
Collecting PyGithub==1.58.1 (from -r E:\GitHub\servo\python\requirements.txt (line 25))
  Using cached PyGithub-1.58.1-py3-none-any.whl.metadata (1.8 kB)
Collecting six==1.16 (from -r E:\GitHub\servo\python\requirements.txt (line 28))
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting notify-py==0.3.42 (from -r E:\GitHub\servo\python\requirements.txt (line 31))
  Using cached notify_py-0.3.42-py3-none-any.whl.metadata (5.2 kB)
Collecting flask (from -r E:\GitHub\servo\python\requirements.txt (line 34))
  Using cached flask-3.0.3-py3-none-any.whl.metadata (3.2 kB)
Collecting requests (from -r E:\GitHub\servo\python\requirements.txt (line 35))
  Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
Collecting types-requests (from -r E:\GitHub\servo\python\requirements.txt (line 36))
  Using cached types_requests-2.32.0.20241016-py3-none-any.whl.metadata (1.9 kB)
Collecting Mako==1.2.2 (from -r E:\GitHub\servo\python\requirements.txt (line 39))
  Using cached Mako-1.2.2-py3-none-any.whl.metadata (2.9 kB)
Collecting httpx==0.27.2 (from httpx[http2]==0.27.2->-r E:\GitHub\servo\tests\wpt\tests\tools\requirements_tests.txt (line 1))
  Using cached httpx-0.27.2-py3-none-any.whl.metadata (7.1 kB)
Collecting json-e==4.7.0 (from -r E:\GitHub\servo\tests\wpt\tests\tools\requirements_tests.txt (line 2))
  Using cached json_e-4.7.0-py3-none-any.whl.metadata (482 bytes)
Collecting jsonschema==4.17.3 (from -r E:\GitHub\servo\tests\wpt\tests\tools\requirements_tests.txt (line 3))
  Using cached jsonschema-4.17.3-py3-none-any.whl.metadata (7.9 kB)
Collecting pyyaml==6.0.1 (from -r E:\GitHub\servo\tests\wpt\tests\tools\requirements_tests.txt (line 4))
  Downloading PyYAML-6.0.1.tar.gz (125 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting taskcluster==68.0.4 (from -r E:\GitHub\servo\tests\wpt\tests\tools\requirements_tests.txt (line 5))
  Using cached taskcluster-68.0.4-py3-none-any.whl.metadata (25 kB)
Collecting mozterm==1.0.0 (from -r E:\GitHub\servo\tests\wpt\tests\tools\requirements_tests.txt (line 6))
  Using cached mozterm-1.0.0-py2.py3-none-any.whl.metadata (669 bytes)
Collecting html5lib==1.1 (from -r E:\GitHub\servo\tests\wpt\tests\tools\wptrunner\requirements.txt (line 1))
  Using cached html5lib-1.1-py2.py3-none-any.whl.metadata (16 kB)
Collecting mozdebug==0.3.1 (from -r E:\GitHub\servo\tests\wpt\tests\tools\wptrunner\requirements.txt (line 2))
  Using cached mozdebug-0.3.1.tar.gz (6.3 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting mozprocess==1.3.1 (from -r E:\GitHub\servo\tests\wpt\tests\tools\wptrunner\requirements.txt (line 5))
  Using cached mozprocess-1.3.1-py2.py3-none-any.whl.metadata (797 bytes)
Collecting packaging==24.0 (from -r E:\GitHub\servo\tests\wpt\tests\tools\wptrunner\requirements.txt (line 6))
  Using cached packaging-24.0-py3-none-any.whl.metadata (3.2 kB)
Collecting pillow==10.3.0 (from -r E:\GitHub\servo\tests\wpt\tests\tools\wptrunner\requirements.txt (line 7))
  Downloading pillow-10.3.0.tar.gz (46.6 MB)
     --------------------------------------- 46.6/46.6 MB 21.7 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'error'
  error: subprocess-exited-with-error

  Getting requirements to build wheel did not run successfully.
  exit code: 1

  [21 lines of output]
  Traceback (most recent call last):
    File "E:\GitHub\servo\python\_venv3.13\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
      main()
      ~~~~^^
    File "E:\GitHub\servo\python\_venv3.13\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
                               ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
    File "E:\GitHub\servo\python\_venv3.13\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 118, in get_requires_for_build_wheel
      return hook(config_settings)
    File "C:\Users\DK\AppData\Local\Temp\pip-build-env-r6t_14du\overlay\Lib\site-packages\setuptools\build_meta.py", line 333, in get_requires_for_build_wheel
      return self._get_build_requires(config_settings, requirements=[])
             ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\[Username]\AppData\Local\Temp\pip-build-env-r6t_14du\overlay\Lib\site-packages\setuptools\build_meta.py", line 303, in _get_build_requires
      self.run_setup()
      ~~~~~~~~~~~~~~^^
    File "C:\Users\[Username]\AppData\Local\Temp\pip-build-env-r6t_14du\overlay\Lib\site-packages\setuptools\build_meta.py", line 319, in run_setup
      exec(code, locals())
      ~~~~^^^^^^^^^^^^^^^^
    File "<string>", line 33, in <module>
    File "<string>", line 27, in get_version
  KeyError: '__version__'
  [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

Getting requirements to build wheel did not run successfully.
exit code: 1

See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

Process failed with return code: 1
```

Using Python 3.12.7 does not lead to the same problem.

**To Reproduce:**
On Windows
- Install Python 3.13.0
- Attempt to run `.\mach bootstrap`

**Platform:**
Tested on freshly installed Windows 11.

### PR DESCRIPTION
**Pull request by MichaelMcDonnell**
The 'cgi' module was removed in Python 3.13. This broke mach bootstrap for Python 13.3. The `legacy-cgi` is a drop-in replacement for projects not ready to upgrade to a better approach. I've added that to the requirements. The newer version of the `pillow` module was also required.

The mach test-wpt command still doesn't work after this change but bootstrap at least works.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #34095

- [X] These changes do not require tests because it can be tested by running `./mach bootstrap` with Python 3.13 installed. We could set up CI to do that if needed.

### CODE REVIEW

**Comment by dklassic** [comment]
Tested on a Windows 11 + Python 3.13.0 setup I can confirm this fix is working!

**Review by wusyong — APPROVED** [review]

**Comment by mrobinson** [comment]
I suspect a change like this should be submitted upstream in the WPT project instead of here so that it can be properly reviewed by the people who maintain the Python dependencies.

Please do not land this change from here.

**Review by mrobinson — CHANGES_REQUESTED** [review]
I have doubts about whether or not this breaks things upstream.

**Comment by MichaelMcDonnell** [comment]
@mrobinson Yes, it would be ideal to make the changes in WPT upstream. Yes, if this change were to be applied in upstream it would break upstream by requiring a Python 3.10 as a minimum instead of Python 3.8. I've run the WPT tests with Python 3.12 and that still worked.

Without this change `./mach build` doesn't work. It means I cannot build Servo on Fedora 41 without installing Python 3.12 and making it the default.

I made this PR as a stop-gap measure to unblock myself and others from building Servo until the issue is fixed upstream. The WPT tests fail inside the Python logging library with a lock being None, which points to an issue in Python itself. I hope that someone else will "pick up the ball" from here and figure out the last changes that are needed to fix the WPT tests. I expect it to be trivial to merge upstream back once it is fixed since they will most likely make the same changes plus some more work.

I guess it comes down to what is prioritized higher, unblocking versus taking on risk of breaking something?

**Comment by MichaelMcDonnell** [comment]
I have created a WIP upstream pull request https://github.com/web-platform-tests/wpt/pull/49503. Help would greatly be appreciated.

**Comment by mrobinson** [comment]
> @mrobinson Yes, it would be ideal to make the changes in WPT upstream. Yes, if this change were to be applied in upstream it would break upstream by requiring a Python 3.10 as a minimum instead of Python 3.8. I've run the WPT tests with Python 3.12 and that still worked.

We automatically sync changes from our repository upstream as upstream trusts us to do reviews here, so merging this here will also merge it upstream. That's the main issue here. Thanks for moving the discussion upstream!

**Comment by mrobinson** [comment]
Closing this one in favor of https://github.com/web-platform-tests/wpt/pull/49503.

**Comment by MichaelMcDonnell** [comment]
> We automatically sync changes from our repository upstream as upstream trusts us to do reviews here, so merging this here will also merge it upstream. That's the main issue here. Thanks for moving the discussion upstream!

Ah ok, yeah that would not work then. Thank you for helping me!

**Comment by MichaelMcDonnell** [comment]
I've closed web-platform-tests/wpt#49503 too. I don't see a way to support both that old versions of Python and current Python.

In the short term, I think the solution is not to rely on the system Python. In the long term, I think moving away from Python and to Rust instead is a better solution.

**Pull request closed without merging**
